presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java

```diff
@@ -511,12 +510,10 @@ public HiveRecordWriter(
             // todo adapt input types to the file types as Hive does
             for (int fileIndex = 0; fileIndex < fileColumnNames.size(); fileIndex++) {
                 String columnName = fileColumnNames.get(fileIndex);
-                Type fileColumnType = fileColumnTypes.get(fileIndex);
-                int inputIndex = inputColumnNames.indexOf(columnName);
-                Type inputType = inputColumnTypes.get(inputIndex);
+                HiveType fileColumnHiveType = fileColumnHiveTypes.get(fileIndex);
+                HiveType inputHiveType = inputColumnMap.get(columnName).getHiveType();
 
-                // HACK: solve coercion in generic way
-                if (!typesMatchForInsert(fileColumnType, inputType)) {
+                if (!fileColumnHiveType.equals(inputHiveType)) {
```

Comment:
@pnowojski How does this line of change affect your VARCHAR changes? Why did you introduce `typesMatchForInsert` given you didn't add Hive Varchar to the list of supported types in `HiveType`